soroban-env-host/tests/bls.rs

```diff
@@ -0,0 +1,449 @@
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct AggregateVerifyInput {
+        pubkeys: Vec<String>,
+        messages: Vec<String>,
+        signature: String,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct AggregateVerify {
+        input: AggregateVerifyInput,
+        output: bool,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct BatchVerifyInput {
+        pubkeys: Vec<String>,
+        messages: Vec<String>,
+        signatures: Vec<String>,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct BatchVerify {
+        input: BatchVerifyInput,
+        output: bool,
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct FastAggregateVerifyInput {
+        pubkeys: Vec<String>,
+        message: String,
+        signature: String,
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct FastAggregateVerify {
+        input: FastAggregateVerifyInput,
+        output: bool,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct VerifyInput {
+        pubkey: String,
+        message: String,
+        signature: String,
+    }
+
+    #[derive(Debug, Deserialize)]
+    struct Verify {
+        input: VerifyInput,
+        output: bool,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct SignInput {
+        privkey: String,
+        message: String,
+    }
+
+    #[derive(Deserialize, Debug)]
+    struct Sign {
+        input: SignInput,
+        output: String,
+    }
+
+    fn parse_hex(s: &str) -> Vec<u8> {
+        Vec::from_hex(s.trim_start_matches("0x")).unwrap()
+    }
+
+    fn adapt_g1_point(host: &Host, p: &String) -> Result<BytesObject, HostError> {
+        let p = parse_hex(p.as_str());
+        assert_eq!(p.len(), 48);
+        let pt =
+            G1Affine::deserialize_with_mode(p.as_slice(), Compress::Yes, Validate::No).unwrap();
+        let mut buf = vec![0u8; 96];
+        pt.serialize_with_mode(buf.as_mut_slice(), Compress::No)
+            .unwrap();
+        host.bytes_new_from_slice(&buf)
+    }
+
+    fn neg_g1(host: &Host) -> Result<BytesObject, HostError> {
+        let mut buf = [0u8; 96];
+        let neg_g1 = -G1Affine::generator();
+        neg_g1.serialize_uncompressed(buf.as_mut_slice()).unwrap();
+        host.bytes_new_from_slice(&buf)
+    }
+
+    fn adapt_g2_point(host: &Host, p: &String) -> Result<BytesObject, HostError> {
+        let p = parse_hex(p.as_str());
+        assert_eq!(p.len(), 96);
+        let pt =
+            G2Affine::deserialize_with_mode(p.as_slice(), Compress::Yes, Validate::No).unwrap();
+        let mut buf = vec![0u8; 192];
+        pt.serialize_with_mode(buf.as_mut_slice(), Compress::No)
+            .unwrap();
+        host.bytes_new_from_slice(&buf)
+    }
+
+    fn hash_msg_to_curve(
+        host: &Host,
+        msg: &String,
+        dst: BytesObject,
+    ) -> Result<BytesObject, HostError> {
+        let msg = parse_hex(msg.as_str());
+        let msg = host.bytes_new_from_slice(&msg)?;
+        host.bls12_381_hash_to_g2(msg, dst)
```

Comment:
Just curious - Is there a reason why the Ethereum tests only use `bls12_381_hash_to_g2` and not `bls12_381_hash_to_g1`?

Reply:
The Ethereum BLS signature maps public key to G1 and message/signature to G2. This is just a convention, and the reason for doing that is to save storage on public keys, since storing G1 requires half the size compared to G2.